src/Compilers/CSharp/Portable/Utilities/ValueSetFactory.NumericValueSet.cs

```diff
@@ -23,72 +26,75 @@ internal static partial class ValueSetFactory
         /// </summary>
         private sealed class NumericValueSet<T, TTC> : IValueSet<T> where TTC : struct, INumericTC<T>
         {
-            private readonly Interval _rootInterval;
+            private readonly ImmutableArray<(T first, T last)> _intervals;
 
-            internal NumericValueSet(Interval rootInterval) => this._rootInterval = rootInterval;
+            public readonly static NumericValueSet<T, TTC> AllValues = new NumericValueSet<T, TTC>(default(TTC).MinValue, default(TTC).MaxValue);
+            public readonly static NumericValueSet<T, TTC> NoValues = new NumericValueSet<T, TTC>(ImmutableArray<(T first, T last)>.Empty);
 
-            bool IValueSet.IsEmpty => _rootInterval is Interval.Excluded;
+            internal NumericValueSet(T first, T last) : this(ImmutableArray.Create((first, last)))
+            {
+                TTC tc = default;
+                Debug.Assert(tc.Related(LessThanOrEqual, first, last));
+            }
+
+            internal NumericValueSet(ImmutableArray<(T first, T last)> intervals)
+            {
+                _intervals = intervals;
+            }
+
+            bool IValueSet.IsEmpty => _intervals.Length == 0;
 
             public bool Any(BinaryOperatorKind relation, T value)
             {
                 TTC tc = default;
-                return anyInterval(_rootInterval, relation, value, tc.MinValue, tc.MaxValue);
+                switch (relation)
+                {
+                    case LessThan:
+                    case LessThanOrEqual:
+                        return _intervals.Length > 0 && tc.Related(relation, _intervals[0].first, value);
+                    case GreaterThan:
+                    case GreaterThanOrEqual:
+                        return _intervals.Length > 0 && tc.Related(relation, _intervals[_intervals.Length - 1].last, value);
+                    case Equal:
+                        return anyIntervalContains(0, _intervals.Length - 1, value);
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(relation);
+                }
 
-                // Compute the result of <see cref="IValueSet{T}.Any"/> for a given interval.
-                static bool anyInterval(Interval interval, BinaryOperatorKind relation, T value, T minValue, T maxValue)
+                bool anyIntervalContains(int firstIntervalIndex, int lastIntervalIndex, T value)
                 {
-                    TTC tc = default;
-                    Debug.Assert(tc.Related(LessThanOrEqual, minValue, maxValue));
-                    switch (interval)
+                    while (true)
                     {
-                        case Interval.Excluded _:
+                        if ((lastIntervalIndex - firstIntervalIndex + 1) < 4)
```

Comment:
Why special case `< 4` rather than relying on the `else` below until `if (lastIntervalIndex < firstIntervalIndex) return false;`?

Reply:
When there are few enough (`< 4`)intervals, we test directly rather than binary search.  The zero intervals situation just falls out as a special case of that.

---
In reply to: [401094613](https://github.com/dotnet/roslyn/pull/42886#discussion_r401094613) [](ancestors = 401094613)